README.md
# Case study


We want to know the __countries that don't host a megapolis__


The purpose of this exercice is to evaluate your skills in Python and SQL. You'll have to fork this repository and write a program that fetch the [dataset of the population of all cities in the world](https://public.opendatasoft.com/explore/dataset/geonames-all-cities-with-a-population-1000/export/?disjunctive.cou_name_en), stores it in a database, then perform a query that will compute: what are the countries that don't host a megapoliss (a city of more than 10,000,000 inhabitants)? 

The program will save the result (country code and country name) as a tabulated separated value file, ordered by country name. 

You should answer as if you were writting production code within your team. You can imagine that the program will be run automatically every week to update the resulting data.

Please send us the link to your github repository with the answer of the exercise. 

# Countries without Megapolis - Case Study Solution by Malak HOUSSAINI

This project provides a solution for the case study on identifying countries without a megapolis. It is implemented in Python using the Pandas library for data manipulation and SQLite for database management.

## Project Overview

The goal of this project is to analyze a dataset of cities and identify countries that do not have a megapolis, i.e., cities with a population over 10 million. The project involves the following steps:

1. **Environment Setup**: Install the required packages (Pandas and SQLAlchemy) by running the command `conda install pandas sqlalchemy` in Anaconda Prompt.

2. **Fork the Repository**: Fork the original repository to create your own copy.

3. **Code Implementation**: Create a new Python script on any IDE (visual studio in my case) and import the necessary libraries: Pandas and SQLAlchemy.

4. **Data Reading**: Read the provided sample dataset (`geonames-all-cities-with-a-population-1000.csv`) into a Pandas DataFrame.

5. **Data Preprocessing**: Filter the dataset to exclude cities with a population over 10 million. Group the data by country code and name and sort it in ascending order of country name.

6. **Save Results**: Save the sorted and filtered data as a tab-separated value (TSV) file named `result.tsv`.

7. **Database Setup**: Establish a connection to a SQLite database (`cities.db`).

8. **Data Storage**: Store the original dataset in the SQLite database for future reference.

9. **Perform SQL Query**: Execute an SQL query on the database to retrieve the countries without a megapolis. The query groups the data by country code and name and orders it by country name.

10. **Save Query Results**: Save the result of the SQL query as a TSV file named `result.tsv`.

11. **Database Cleanup**: Close the database connection.

12. **Commit and Push**: Commit your code changes and push them to your GitHub repository.

## Running the Project

To run the project, follow these steps:

1. Set up the environment(on anaconda) and install the required packages using the provided command.

2. Fork the repository to create your own copy.

3. Create a new Python script and import the required libraries.

4. Use the code implementation steps mentioned above into your script.

5. Place the `geonames-all-cities-with-a-population-1000.csv` file in the same directory as your script.

6. Run the script.

7. The script will generate the `result.tsv` file, which contains the list of countries without a megapolis.

Deployment:
- This project is intended to be run locally.
- To automate it, you can set up a scheduled task or use the Task Scheduler on your operating system to run the Python scripts periodically, or once in a week to update the result data.

Built With:
- Python 3.x (3.10.11): The programming language used.
- SQLite: The database engine used for storing and querying the data.

Author:
- [Malak HOUSSAINI](https://github.com/malakhousseiny)

License:
- This project is licensed under the MIT License.


conda-meta/history
==> 2023-06-11 13:21:34 <==
# cmd: C:\Users\HMala\anaconda3\Scripts\conda-script.py create --yes --json --prefix C:\Users\HMala\anaconda3\envs\CaseStudy1 python=3.10
# conda version: 23.3.1
+defaults/noarch::tzdata-2023c-h04d1e81_0
+defaults/win-64::bzip2-1.0.8-he774522_0
+defaults/win-64::ca-certificates-2023.05.30-haa95532_0
+defaults/win-64::libffi-3.4.4-hd77b12b_0
+defaults/win-64::openssl-3.0.8-h2bbff1b_0
+defaults/win-64::pip-23.0.1-py310haa95532_0
+defaults/win-64::python-3.10.11-he1021f5_3
+defaults/win-64::setuptools-67.8.0-py310haa95532_0
+defaults/win-64::sqlite-3.41.2-h2bbff1b_0
+defaults/win-64::tk-8.6.12-h2bbff1b_0
+defaults/win-64::vc-14.2-h21ff451_1
+defaults/win-64::vs2015_runtime-14.27.29016-h5e58377_2
+defaults/win-64::wheel-0.38.4-py310haa95532_0
+defaults/win-64::xz-5.4.2-h8cc25b3_0
+defaults/win-64::zlib-1.2.13-h8cc25b3_0
# update specs: ['python=3.10']
==> 2023-06-11 13:23:32 <==
# cmd: C:\Users\HMala\anaconda3\Scripts\conda-script.py install --yes --json --force-pscheck --prefix C:\Users\HMala\anaconda3\envs\CaseStudy1 pandas sqlalchemy
# conda version: 23.3.1
+defaults/noarch::python-dateutil-2.8.2-pyhd3eb1b0_0
+defaults/noarch::six-1.16.0-pyhd3eb1b0_1
+defaults/win-64::blas-1.0-mkl
+defaults/win-64::bottleneck-1.3.5-py310h9128911_0
+defaults/win-64::greenlet-2.0.1-py310hd77b12b_0
+defaults/win-64::intel-openmp-2023.1.0-h59b6b97_46319
+defaults/win-64::mkl-2023.1.0-h8bd8f75_46356
+defaults/win-64::mkl-service-2.4.0-py310h2bbff1b_1
+defaults/win-64::mkl_fft-1.3.6-py310h4ed8f06_1
+defaults/win-64::mkl_random-1.2.2-py310h4ed8f06_1
+defaults/win-64::numexpr-2.8.4-py310h2cd9be0_1
+defaults/win-64::numpy-1.24.3-py310h055cbcc_1
+defaults/win-64::numpy-base-1.24.3-py310h65a83cf_1
+defaults/win-64::pandas-1.5.3-py310h4ed8f06_0
+defaults/win-64::pytz-2022.7-py310haa95532_0
+defaults/win-64::sqlalchemy-1.4.39-py310h2bbff1b_0
+defaults/win-64::tbb-2021.8.0-h59b6b97_0
# update specs: ['sqlalchemy', 'pandas']


Library/lib/tclConfig.sh
# tclConfig.sh --
#
# This shell script (for sh) is generated automatically by Tcl's
# configure script.  It will create shell variables for most of
# the configuration options discovered by the configure script.
# This script is intended to be included by the configure scripts
# for Tcl extensions so that they don't have to figure this all
# out for themselves.
#
# The information in this file is specific to a single platform.

TCL_DLL_FILE="tcl86t.dll"

# Tcl's version number.
TCL_VERSION='8.6'
TCL_MAJOR_VERSION='8'
TCL_MINOR_VERSION='6'
TCL_PATCH_LEVEL='8.6.12'

# C compiler to use for compilation.
TCL_CC='cl.exe'

# -D flags for use with the C compiler.
TCL_DEFS='-nologo -c /D_ATL_XP_TARGETING  -W3 -wd4311 -wd4312 -FpC:\ci\tk_1654036608655\work\tcl8.6.12\win\Release_AMD64_VC1916\tcl_ThreadedDynamic\  -fp:strict -O2 -GS -GL -MD -I"C:\ci\tk_1654036608655\work\tcl8.6.12\win\..\win" -I"C:\ci\tk_1654036608655\work\tcl8.6.12\win\..\generic"  -I"C:\ci\tk_1654036608655\work\tcl8.6.12\win\..\libtommath"  /DTCL_TOMMATH /DMP_PREC=4 /Dinline=__inline /DHAVE_ZLIB=1 /D_CRT_SECURE_NO_DEPRECATE /D_CRT_NONSTDC_NO_DEPRECATE /DMP_FIXED_CUTOFFS /DSTDC_HEADERS /DUSE_NMAKE=1 /DHAVE_STDINT_H=1 /DHAVE_INTTYPES_H=1 /DHAVE_STDBOOL_H=1 /DTCL_THREADS=1 /DUSE_THREAD_ALLOC=1 /DNDEBUG /DTCL_CFG_OPTIMIZED /DTCL_CFG_DO64BIT   /DBUILD_tcl'

# If TCL was built with debugging symbols, generated libraries contain
# this string at the end of the library name (before the extension).
TCL_DBGX=t

# Default flags used in an optimized and debuggable build, respectively.
TCL_CFLAGS_DEBUG='-nologo -c -W3 -YX -FpC:\ci\tk_1654036608655\work\tcl8.6.12\win\Release_AMD64_VC1916\tcl_ThreadedDynamic\ -MDd'
TCL_CFLAGS_OPTIMIZE='-nologo -c -W3 -YX -FpC:\ci\tk_1654036608655\work\tcl8.6.12\win\Release_AMD64_VC1916\tcl_ThreadedDynamic\ -MD'

# Default linker flags used in an optimized and debuggable build, respectively.
TCL_LDFLAGS_DEBUG='-nologo -machine:AMD64 -debug -debugtype:cv'
TCL_LDFLAGS_OPTIMIZE='-nologo -machine:AMD64 -release -opt:ref -opt:icf,3'

# Flag, 1: we built a shared lib, 0 we didn't
TCL_SHARED_BUILD=1

# The name of the Tcl library (may be either a .a file or a shared library):
TCL_LIB_FILE='tcl86t.lib'

# Flag to indicate whether shared libraries need export files.
TCL_NEEDS_EXP_FILE=

# String that can be evaluated to generate the part of the export file
# name that comes after the "libxxx" (includes version number, if any,
# extension, and anything else needed).  May depend on the variables
# VERSION.  On most UNIX systems this is ${VERSION}.exp.
TCL_EXPORT_FILE_SUFFIX='86t.lib'

# Additional libraries to use when linking Tcl.
TCL_LIBS='kernel32.lib advapi32.lib netapi32.lib user32.lib userenv.lib ws2_32.lib ucrt.lib netapi32.lib user32.lib userenv.lib ws2_32.lib'

# Top-level directory in which Tcl's platform-independent files are
# installed.
TCL_PREFIX='C:/Users/HMala/anaconda3/envs/CaseStudy1\Library'

# Top-level directory in which Tcl's platform-specific files (e.g.
# executables) are installed.
TCL_EXEC_PREFIX='C:/Users/HMala/anaconda3/envs/CaseStudy1\Library\bin'

# Flags to pass to cc when compiling the components of a shared library:
TCL_SHLIB_CFLAGS=''

# Flags to pass to cc to get warning messages
TCL_CFLAGS_WARNING='-W3'

# Extra flags to pass to cc:
TCL_EXTRA_CFLAGS='-YX'

# Base command to use for combining object files into a shared library:
TCL_SHLIB_LD='link -nologo -machine:AMD64  -ltcg -release -opt:ref -opt:icf,3 -nodefaultlib:libucrt.lib -dll'

# Base command to use for combining object files into a static library:
TCL_STLIB_LD='lib -nologo'

# Either '$LIBS' (if dependent libraries should be included when linking
# shared libraries) or an empty string.  See Tcl's configure.in for more
# explanation.
TCL_SHLIB_LD_LIBS='kernel32.lib advapi32.lib netapi32.lib user32.lib userenv.lib ws2_32.lib ucrt.lib netapi32.lib user32.lib userenv.lib ws2_32.lib'

# Suffix to use for the name of a shared library.
TCL_SHLIB_SUFFIX='.dll'

# Library file(s) to include in tclsh and other base applications
# in order to provide facilities needed by DLOBJ above.
TCL_DL_LIBS=''

# Flags to pass to the compiler when linking object files into
# an executable tclsh or tcltest binary.
TCL_LD_FLAGS=''

# Flags to pass to cc/ld, such as "-R /usr/local/tcl/lib", that tell the
# run-time dynamic linker where to look for shared libraries such as
# libtcl.so.  Used when linking applications.  Only works if there
# is a variable "LIB_RUNTIME_DIR" defined in the Makefile.
TCL_CC_SEARCH_FLAGS=''
TCL_LD_SEARCH_FLAGS=''

# Additional object files linked with Tcl to provide compatibility
# with standard facilities from ANSI C or POSIX.
TCL_COMPAT_OBJS=''

# Name of the ranlib program to use.
TCL_RANLIB=''

# -l flag to pass to the linker to pick up the Tcl library
TCL_LIB_FLAG=''

# String to pass to linker to pick up the Tcl library from its
# build directory.
TCL_BUILD_LIB_SPEC=''

# String to pass to linker to pick up the Tcl library from its
# installed directory.
TCL_LIB_SPEC='C:/Users/HMala/anaconda3/envs/CaseStudy1\Library\lib\tcl86t.lib'

# String to pass to the compiler so that an extension can
# find installed Tcl headers.
TCL_INCLUDE_SPEC='-IC:/Users/HMala/anaconda3/envs/CaseStudy1\Library\include'

# Indicates whether a version numbers should be used in -l switches
# ("ok" means it's safe to use switches like -ltcl7.5;  "nodots" means
# use switches like -ltcl75).  SunOS and FreeBSD require "nodots", for
# example.
TCL_LIB_VERSIONS_OK=''

# String that can be evaluated to generate the part of a shared library
# name that comes after the "libxxx" (includes version number, if any,
# extension, and anything else needed).  May depend on the variables
# VERSION and SHLIB_SUFFIX.  On most UNIX systems this is
# ${VERSION}${SHLIB_SUFFIX}.
TCL_SHARED_LIB_SUFFIX='86t.dll'

# String that can be evaluated to generate the part of an unshared library
# name that comes after the "libxxx" (includes version number, if any,
# extension, and anything else needed).  May depend on the variable
# VERSION.  On most UNIX systems this is ${VERSION}.a.
TCL_UNSHARED_LIB_SUFFIX='86t.lib'

# Location of the top-level source directory from which Tcl was built.
# This is the directory that contains a README file as well as
# subdirectories such as generic, unix, etc.  If Tcl was compiled in a
# different place than the directory containing the source files, this
# points to the location of the sources, not the location where Tcl was
# compiled.
TCL_SRC_DIR='C:\ci\tk_1654036608655\work\tcl8.6.12\win\..'

# List of standard directories in which to look for packages during
# "package require" commands.  Contains the "prefix" directory plus also
# the "exec_prefix" directory, if it is different.
TCL_PACKAGE_PATH=''

# Tcl supports stub.
TCL_SUPPORTS_STUBS=1

# The name of the Tcl stub library (.a):
TCL_STUB_LIB_FILE='tclstub86.lib'

# -l flag to pass to the linker to pick up the Tcl stub library
TCL_STUB_LIB_FLAG='tclstub86.lib'

# String to pass to linker to pick up the Tcl stub library from its
# build directory.
TCL_BUILD_STUB_LIB_SPEC='-LC:\ci\tk_1654036608655\work\tcl8.6.12\win\Release_AMD64_VC1916 tclstub86.lib'

# String to pass to linker to pick up the Tcl stub library from its
# installed directory.
TCL_STUB_LIB_SPEC='-LC:/Users/HMala/anaconda3/envs/CaseStudy1\Library\lib tclstub86.lib'

# Path to the Tcl stub library in the build directory.
TCL_BUILD_STUB_LIB_PATH='C:\ci\tk_1654036608655\work\tcl8.6.12\win\Release_AMD64_VC1916\tclstub86.lib'

# Path to the Tcl stub library in the install directory.
TCL_STUB_LIB_PATH='C:/Users/HMala/anaconda3/envs/CaseStudy1\Library\lib\tclstub86.lib'

# Flag, 1: we built Tcl with threads enabled, 0 we didn't
TCL_THREADS=1



Library/lib/tdbc1.1.3/tdbcConfig.sh
# tdbcConfig.sh --
#
# This shell script (for sh) is generated automatically by TDBC's configure
# script. It will create shell variables for most of the configuration options
# discovered by the configure script. This script is intended to be included
# by the configure scripts for TDBC extensions so that they don't have to
# figure this all out for themselves.
#
# The information in this file is specific to a single platform.
#
# RCS: @(#) $Id$

# TDBC's version number
tdbc_VERSION=1.1.3
TDBC_VERSION=1.1.3

# Name of the TDBC library - may be either a static or shared library
tdbc_LIB_FILE=tdbc113t.dll
TDBC_LIB_FILE=tdbc113t.dll

# String to pass to the linker to pick up the TDBC library from its build dir
tdbc_BUILD_LIB_SPEC="@tdbc_BUILD_LIB_SPEC@"
TDBC_BUILD_LIB_SPEC="@tdbc_BUILD_LIB_SPEC@"

# String to pass to the linker to pick up the TDBC library from its installed
# dir.
tdbc_LIB_SPEC="C:/Users/HMala/anaconda3/envs/CaseStudy1\Library\lib\tdbc1.1.3\tdbc113t.dll"
TDBC_LIB_SPEC="C:/Users/HMala/anaconda3/envs/CaseStudy1\Library\lib\tdbc1.1.3\tdbc113t.dll"

# Name of the TBDC stub library
tdbc_STUB_LIB_FILE="tdbcstub113.lib"
TDBC_STUB_LIB_FILE="tdbcstub113.lib"

# String to pass to the linker to pick up the TDBC stub library from its
# build directory
tdbc_BUILD_STUB_LIB_SPEC="C:\ci\tk_1654036608655\work\tcl8.6.12\pkgs\tdbc1.1.3\win\Release_AMD64_VC1916\tdbcstub113.lib"
TDBC_BUILD_STUB_LIB_SPEC="C:\ci\tk_1654036608655\work\tcl8.6.12\pkgs\tdbc1.1.3\win\Release_AMD64_VC1916\tdbcstub113.lib"

# String to pass to the linker to pick up the TDBC stub library from its
# installed directory
tdbc_STUB_LIB_SPEC="C:/Users/HMala/anaconda3/envs/CaseStudy1\Library\lib\tdbc1.1.3\tdbcstub113.lib"
TDBC_STUB_LIB_SPEC="C:/Users/HMala/anaconda3/envs/CaseStudy1\Library\lib\tdbc1.1.3\tdbcstub113.lib"

# Path name of the TDBC stub library in its build directory
tdbc_BUILD_STUB_LIB_PATH="C:\ci\tk_1654036608655\work\tcl8.6.12\pkgs\tdbc1.1.3\win\Release_AMD64_VC1916\tdbcstub113.lib"
TDBC_BUILD_STUB_LIB_PATH="C:\ci\tk_1654036608655\work\tcl8.6.12\pkgs\tdbc1.1.3\win\Release_AMD64_VC1916\tdbcstub113.lib"

# Path name of the TDBC stub library in its installed directory
tdbc_STUB_LIB_PATH="C:/Users/HMala/anaconda3/envs/CaseStudy1\Library\lib\tdbc1.1.3\tdbcstub113.lib"
TDBC_STUB_LIB_PATH="C:/Users/HMala/anaconda3/envs/CaseStudy1\Library\lib\tdbc1.1.3\tdbcstub113.lib"

# Location of the top-level source directories from which TDBC was built.
# This is the directory that contains doc/, generic/ and so on.  If TDBC
# was compiled in a directory other than the source directory, this still
# points to the location of the sources, not the location where TDBC was
# compiled.
tdbc_SRC_DIR="C:\ci\tk_1654036608655\work\tcl8.6.12\pkgs\tdbc1.1.3\win\.."
TDBC_SRC_DIR="C:\ci\tk_1654036608655\work\tcl8.6.12\pkgs\tdbc1.1.3\win\.."

# String to pass to the compiler so that an extension can find installed TDBC
# headers
tdbc_INCLUDE_SPEC="-IC:/Users/HMala/anaconda3/envs/CaseStudy1\Library\lib\..\include"
TDBC_INCLUDE_SPEC="-IC:/Users/HMala/anaconda3/envs/CaseStudy1\Library\lib\..\include"

# String to pass to the compiler so that an extension can find TDBC headers
# in the source directory
tdbc_BUILD_INCLUDE_SPEC="-IC:\ci\tk_1654036608655\work\tcl8.6.12\pkgs\tdbc1.1.3\win\..\generic"
TDBC_BUILD_INCLUDE_SPEC="-IC:\ci\tk_1654036608655\work\tcl8.6.12\pkgs\tdbc1.1.3\win\..\generic"

# Path name where .tcl files in the tdbc package appear at run time.
tdbc_LIBRARY_PATH="C:/Users/HMala/anaconda3/envs/CaseStudy1\Library\lib\tdbc1.1.3"
TDBC_LIBRARY_PATH="C:/Users/HMala/anaconda3/envs/CaseStudy1\Library\lib\tdbc1.1.3"

# Path name where .tcl files in the tdbc package appear at build time.
tdbc_BUILD_LIBRARY_PATH="C:\ci\tk_1654036608655\work\tcl8.6.12\pkgs\tdbc1.1.3\win\../library"
TDBC_BUILD_LIBRARY_PATH="C:\ci\tk_1654036608655\work\tcl8.6.12\pkgs\tdbc1.1.3\win\../library"

# Additional flags that must be passed to the C compiler to use tdbc
tdbc_CFLAGS=
TDBC_CFLAGS=



casestudy1-megapolis.py
import pandas as pd
from sqlalchemy import create_engine


#to read the given data set
data = pd.read_csv('geonames-all-cities-with-a-population-1000.csv', sep=';', encoding='latin1')


#To Preprocess the dataset,selecting only the following columns: 'Country Code', 'Country name EN', and 'Population'
#filter the cities whose population > 10,000,000 to remove out of the new dataset
#group the data by 'country code' and 'country name EN'as implied by the case study to get countries that don't host a megapolis
#Sort the data by 'Country name EN'
filtered_data = data[data['Population'] <= 10000000]
grouped_data = filtered_data.groupby(['Country Code', 'Country name EN'], as_index=False).size()
sorted_data = grouped_data.sort_values('Country name EN')

#save the result as a tab-separated value (TSV) file names 'result.tsv'
result = sorted_data[['Country Code', 'Country name EN']]
result.to_csv('result.tsv', sep='\t', index=False)

#setting up a database connection 
engine = create_engine('sqlite:///cities.db')
connection = engine.connect()

#store the dataset in the database
data.to_sql('cities', con=connection, if_exists='replace')

#Perform a query on SQL to retrieve countries that don't host a megapolis
query = '''
SELECT "Country Code", "Country name EN"
FROM cities
WHERE Population <= 10000000
GROUP BY "Country Code", "Country name EN"
ORDER BY "Country name EN"
'''
result = pd.read_sql_query(query, connection)

#Save the SQL query result as a TSV file:
result.to_csv('result.tsv', sep='\t', index=False)

#Close the database connection:
connection.close()

#the result.tsv include some country codes without their corresponding country names as the names were not given in the original dataset 